### ISSUE TITLE
Rule idea: no unknown method binding

### ISSUE THREAD

**Issue by jiayihu** [yannickcr/eslint-plugin-react#868]
The rule should warn when trying to bind a method which doesn't exist on the component class. For example:

``` javascript
export default class MyComponent extends React.Component {
  constructor() {
    super();

    this.unknownMethod = this.unknownMethod.bind(this);
  }
}
```

This bug causes React to throw an error with an unrelated message and difficult to debug.

![schermata 2016-09-28 alle 20 26 11](https://cloud.githubusercontent.com/assets/10067273/18926955/06b24502-85ba-11e6-96b9-ac0d1be80b2d.png)
(Note that in the screenshot AppContainer is the root Component, not the one which binded the unknown method, making it very hard to debug.)

**Comment by ljharb** [comment]
You can't statically know that the method doesn't exist, because you have no idea what `React.Component.prototype` implements.

**Comment by jiayihu** [comment]
Most of the time, apart from lifecycle methods, you don't use any another method from `React.Component.prototype`. Which are the methods, other than lifecycle and `render`, which you could use from React.Component?

**Comment by ljharb** [comment]
What I mean is, this would be a weird React-specific rule, and could only apply to things directly inheriting from `React.Component` - it would make more sense as a generic `class` rule, but then it becomes impossible to statically enforce.

**Comment by lencioni** [comment]
I think there is benefit in having a React-specific rule for this.

**Comment by jiayihu** [comment]
@ljharb I understand your point and the reason I opened the issue here is indeed for the same reason, which is the impossibility to use it with a generic `class`. As I said this rule would apply only to React Components and I don't believe it's weird because binding in constructor follows from the `jsx-no-bind` rule.

**Comment by ljharb** [comment]
I'm concerned that it would break in a future version of React, but we could certainly implement it for 0.13, 0.14, and 15.0
